File: workspace_js/services/js-workflow-executor/src/bootstrap/cache.ts
import type { ConfigModule, Logger } from '@mapexos/microservices';

import { container } from 'tsyringe';

import { New as NewMinIO, NewTieredCache } from '@mapexos/infrastructure';

// InitCache registers MinIO client and TieredCache instances (script source, bytecode) in DI container.
/** Initializes TieredCache instances (script source + bytecode) and MinIO client. */
export async function initCache(configModule: ConfigModule, logger: Logger) {

	// MinIO base configuration
	const minioEndpoint = configModule.get('minio_endpoint') as string;
	const [minioHost, minioPortStr] = minioEndpoint.split(':');
	const minioPort = parseInt(minioPortStr) || 9000;

	const minioBaseConfig = {
		Endpoint: minioHost,
		Port: minioPort,
		AccessKeyID: configModule.get('minio_access_key') as string,
		SecretAccessKey: configModule.get('minio_secret_key') as string,
		UseSSL: configModule.get('minio_use_ssl') as boolean,
		Region: configModule.get('minio_region') as string,
	};

	// TieredCache shared configuration (L0 + L1)
	const serviceName = configModule.get('service_name') as string;
	
	const cacheL0MaxSize = configModule.get('cache_l0_max_size') as number;
	const cacheL0MaxItems = configModule.get('cache_l0_max_items') as number;
	const cacheL0TTL = (configModule.get('cache_l0_ttl_seconds') as number) * 1000;
	
	const cacheL1Enabled = configModule.get('cache_l1_enabled') as boolean;
	const cacheL1BaseDir = configModule.get('cache_l1_dir') as string;
	const cacheL1Dir = `${cacheL1BaseDir}/${serviceName}`;
	const cacheL1MaxSize = configModule.get('cache_l1_max_size') as number;
	const cacheL1TTL = (configModule.get('cache_l1_ttl_seconds') as number) * 1000;

	// Fallback configuration — calls Workflow Service when L2 misses
	const workflowServiceURL = configModule.get('workflow_service_url') as string;
	const internalAPIKey = configModule.get('internal_api_key') as string;
	const fallbackTimeout = configModule.get('cache_fallback_timeout') as number;

	/** MinIO Client for Workflows (single bucket: mapex-workflows) */
	// Path structure: {orgId}/{workflowId}/scripts/{nodeId}.json (source)
	//                 {orgId}/{workflowId}/bytecode/{nodeId}.bin (compiled)
	const minioWorkflowsClient = await NewMinIO({
		...minioBaseConfig,
		BucketName: configModule.get('minio_workflows_bucket') as string,
	});

	container.register('MinIOWorkflowsClient', { useValue: minioWorkflowsClient });
	logger.info('[APP:BOOTSTRAP] MinIO Workflows client initialized');

	/** TieredCache for Script Source (L0 + L1 + L2 + Fallback) */
	// Key format: {orgId}/{workflowId}/scripts/{nodeId}
	// L2 fetches from MinIO: {orgId}/{workflowId}/scripts/{nodeId}.json
	// Fallback HTTP: GET /internal/definitions/{orgId}/{workflowId}/nodes/{nodeId}/script
	const scriptSourceCache = NewTieredCache({
		EnableL0: true,
		L0MaxSize: cacheL0MaxSize,
		L0MaxItems: cacheL0MaxItems,
		L0DefaultTTL: cacheL0TTL,

		EnableL1: cacheL1Enabled,
		L1Dir: cacheL1Dir + '/scripts',
		L1MaxSize: cacheL1MaxSize,
		L1DefaultTTL: cacheL1TTL,

		// L2 loader for script source (MinIO)
		EnableL2: true,
		L2Loader: async (key: string) => {
			try {
				const result = await minioWorkflowsClient.Get(key + '.json');
				return result.Data;
			} catch {
				return null;
			}
		},

		// Fallback: call Workflow Service HTTP endpoint
		FallbackBaseURL: workflowServiceURL,
		FallbackAPIKey: internalAPIKey,
		FallbackEndpoint: '/internal/workflow-scripts',
		FallbackTimeout: fallbackTimeout,
		
		FallbackKeyTransformer: (key: string) => {
			// key format: "{orgId}/{defId}/scripts/{nodeId}"
			// URL path: "/{defId}/scripts/{nodeId}" (strip orgId prefix)
			const firstSlash = key.indexOf('/');
			return firstSlash !== -1 ? key.substring(firstSlash) : `/${key}`;
		},
	});

	container.register('ScriptSourceCache', { useValue: scriptSourceCache });
	logger.info({ l1: cacheL1Enabled, fallback: workflowServiceURL }, '[APP:BOOTSTRAP] TieredCache (script source) initialized');

	/** TieredCache for Bytecode (L1 + L2, no fallback) */
	// L0 (RAM) is SKIPPED by default since piscina workers cache compiled Scripts in-memory.
	// L2 (MinIO) benefits horizontal scaling: Pod A compiles → Pod B reuses.
	// Key format: {orgId}/{workflowId}/bytecode/{nodeId}
	const skipBytecodeL0 = configModule.get('cache_bytecode_skip_l0') as boolean;
	const bytecodeCache = NewTieredCache({
		EnableL0: !skipBytecodeL0,
		L0MaxSize: cacheL0MaxSize,
		L0MaxItems: cacheL0MaxItems,
		L0DefaultTTL: cacheL0TTL,

		EnableL1: cacheL1Enabled,
		L1Dir: cacheL1Dir + '/bytecode',
		L1MaxSize: cacheL1MaxSize,
		L1DefaultTTL: cacheL1TTL,

		// L2 loader for bytecode (MinIO)
		EnableL2: true,
		L2Loader: async (key: string) => {
			try {
				const result = await minioWorkflowsClient.Get(key + '.bin');
				return result.Data;
			} catch {
				return null;
			}
		},
	});

	container.register('BytecodeCache', { useValue: bytecodeCache });
	logger.info({ l1: cacheL1Enabled }, '[APP:BOOTSTRAP] TieredCache (bytecode) initialized');
}

File: workspace_go/services/router/src/bootstrap/cache.go
package bootstrap

import (
	"context"
	"time"

	"go.uber.org/dig"

	config "github.com/Mapex-Solutions/mapexGoKit/microservices/config"
	container "github.com/Mapex-Solutions/mapexGoKit/microservices/container"
	logger "github.com/Mapex-Solutions/mapexGoKit/microservices/logger"

	common "github.com/Mapex-Solutions/mapexGoKit/infrastructure/common/ports"
	minioModel "github.com/Mapex-Solutions/mapexGoKit/infrastructure/minio"
	tieredCacheModel "github.com/Mapex-Solutions/mapexGoKit/infrastructure/tieredcache"
)

// InitTieredCache registers MinIO clients and TieredCaches for assets and templates.
// Both caches use the same L0/L1/L2/Fallback architecture with different MinIO buckets.
func InitTieredCache(c *dig.Container) {
	// Initialize MinIO Client for Assets (L2 Cache - TieredCache)
	minioEndpoint, _ := config.GetStringValue("minio_endpoint")
	minioAccessKey, _ := config.GetStringValue("minio_access_key")
	minioSecretKey, _ := config.GetStringValue("minio_secret_key")
	minioUseSSL := config.GetConfigValue("minio_use_ssl").(bool)
	minioRegion, _ := config.GetStringValue("minio_region")
	minioAssetsBucket, _ := config.GetStringValue("minio_assets_bucket")

	c.Provide(func() *minioModel.MinIOClient {
		mc, err := minioModel.New(minioModel.Config{
			Endpoint:        minioEndpoint,
			AccessKeyID:     minioAccessKey,
			SecretAccessKey: minioSecretKey,
			UseSSL:          minioUseSSL,
			Region:          minioRegion,
			BucketName:      minioAssetsBucket,
			KeyPrefix:       "assets/", // Must match assets service write prefix
		})
		if err != nil {
			logger.Panic("Failed to connect to MinIO: " + err.Error())
		}
		logger.Info("[APP:BOOTSTRAP] MinIO Assets client initialized (bucket: " + minioAssetsBucket + ")")
		return mc
	}, container.Name("assets"))

	// Initialize TieredCache for Assets (L0 + L1 + L2 + Fallback)
	c.Provide(func(params struct {
		container.In
		MinIOClient *minioModel.MinIOClient `name:"assets"`
	}) common.TieredCache {
		
		cacheL0MaxSize, _ := config.GetIntValue("cache_l0_max_size")
		cacheL0MaxItems, _ := config.GetIntValue("cache_l0_max_items")
		cacheL0TTL, _ := config.GetIntValue("cache_l0_ttl_seconds")
		
		cacheL1BaseDir, _ := config.GetStringValue("cache_l1_dir")
		serviceName, _ := config.GetStringValue("service_name")
		cacheL1Dir := cacheL1BaseDir + "/" + serviceName // Include service name to avoid conflicts
		cacheL1MaxSize, _ := config.GetIntValue("cache_l1_max_size")
		cacheL1TTL, _ := config.GetIntValue("cache_l1_ttl_seconds")

		// Fallback configuration - calls Assets Service when L2 misses
		assetsURL, _ := config.GetStringValue("assets_url")
		internalAPIKey, _ := config.GetStringValue("internal_api_key")
		fallbackTimeout, _ := config.GetIntValue("cache_fallback_timeout")

		tc, err := tieredCacheModel.New(tieredCacheModel.Config{
			EnableL0:     true,
			L0MaxSize:    int64(cacheL0MaxSize),
			L0MaxItems:   int64(cacheL0MaxItems),
			L0DefaultTTL: time.Duration(cacheL0TTL) * time.Second,
			EnableL1:     true,
			L1Dir:        cacheL1Dir + "/assets",
			L1MaxSize:    int64(cacheL1MaxSize),
			L1DefaultTTL: time.Duration(cacheL1TTL) * time.Second,

			// L2 loader to fetch from MinIO
			// Key format: {orgId}/{assetUUID} → MinIO path: {orgId}/{assetUUID}.json
			EnableL2: true,
			L2Loader: func(ctx context.Context, key string) ([]byte, error) {
				result, err := params.MinIOClient.Get(ctx, key+".json")
				if err != nil {
					return nil, err
				}
				return result.Data, nil
			},

			// Fallback configuration
			FallbackBaseURL:  assetsURL,
			FallbackAPIKey:   internalAPIKey,
			FallbackEndpoint: "/internal/assets",
			FallbackTimeout:  time.Duration(fallbackTimeout) * time.Second,
		})
		if err != nil {
			logger.Panic("Failed to initialize TieredCache (assets): " + err.Error())
		}

		logger.Info("[APP:BOOTSTRAP] TieredCache (assets) initialized (L0: true, L1: true, Fallback: " + assetsURL + ")")
		return tc
	}, container.Name("assets"))

	/**
	 * Template TieredCache (L0 + L1 + L2 MinIO + Fallback HTTP)
	 * Used by EventService to enrich events with template name and description.
	 * Cache key format: {templateOrgId}/{templateId}
	 * MinIO path: templates/{templateOrgId}/{templateId}.json
	 */

	minioTemplatesBucket, _ := config.GetStringValue("minio_templates_bucket")

	// MinIO Client for Templates (L2 Cache source)
	c.Provide(func() *minioModel.MinIOClient {
		mc, err := minioModel.New(minioModel.Config{
			Endpoint:        minioEndpoint,
			AccessKeyID:     minioAccessKey,
			SecretAccessKey: minioSecretKey,
			UseSSL:          minioUseSSL,
			Region:          minioRegion,
			BucketName:      minioTemplatesBucket,
			KeyPrefix:       "templates/",
		})
		if err != nil {
			logger.Panic("Failed to connect to MinIO (templates): " + err.Error())
		}
		logger.Info("[APP:BOOTSTRAP] MinIO Templates client initialized (bucket: " + minioTemplatesBucket + ")")
		return mc
	}, container.Name("templates"))

	// TieredCache for Templates (reuses same L0/L1 config as assets)
	c.Provide(func(params struct {
		container.In
		MinIOClient *minioModel.MinIOClient `name:"templates"`
	}) common.TieredCache {
		cacheL0MaxSize, _ := config.GetIntValue("cache_l0_max_size")
		cacheL0MaxItems, _ := config.GetIntValue("cache_l0_max_items")
		cacheL0TTL, _ := config.GetIntValue("cache_l0_ttl_seconds")

		cacheL1BaseDir, _ := config.GetStringValue("cache_l1_dir")
		serviceName, _ := config.GetStringValue("service_name")
		cacheL1Dir := cacheL1BaseDir + "/" + serviceName
		cacheL1MaxSize, _ := config.GetIntValue("cache_l1_max_size")
		cacheL1TTL, _ := config.GetIntValue("cache_l1_ttl_seconds")

		assetsURL, _ := config.GetStringValue("assets_url")
		internalAPIKey, _ := config.GetStringValue("internal_api_key")
		fallbackTimeout, _ := config.GetIntValue("cache_fallback_timeout")

		tc, err := tieredCacheModel.New(tieredCacheModel.Config{
			EnableL0:     true,
			L0MaxSize:    int64(cacheL0MaxSize),
			L0MaxItems:   int64(cacheL0MaxItems),
			L0DefaultTTL: time.Duration(cacheL0TTL) * time.Second,
			EnableL1:     true,
			L1Dir:        cacheL1Dir + "/templates",
			L1MaxSize:    int64(cacheL1MaxSize),
			L1DefaultTTL: time.Duration(cacheL1TTL) * time.Second,

			// L2 loader — fetches from MinIO
			// Key format: {templateOrgId}/{templateId} → MinIO path: {templateOrgId}/{templateId}.json
			EnableL2: true,
			L2Loader: func(ctx context.Context, key string) ([]byte, error) {
				result, err := params.MinIOClient.Get(ctx, key+".json")
				if err != nil {
					return nil, err
				}
				return result.Data, nil
			},

			// Fallback — calls Assets Service /internal/templates endpoint
			FallbackBaseURL:  assetsURL,
			FallbackAPIKey:   internalAPIKey,
			FallbackEndpoint: "/internal/templates",
			FallbackTimeout:  time.Duration(fallbackTimeout) * time.Second,
		})
		if err != nil {
			logger.Panic("Failed to initialize TieredCache (templates): " + err.Error())
		}

		logger.Info("[APP:BOOTSTRAP] TieredCache (templates) initialized (L0: true, L1: true, Fallback: " + assetsURL + ")")
		return tc
	}, container.Name("templates"))
}
